code/socialDistribution/views.py

```diff
@@ -384,21 +397,41 @@ def editPost(request, id):
                 post.unlisted = form.cleaned_data.get('unlisted')
                 post.content_media = content_media
 
-                categories = form.cleaned_data.get('categories').split()
-                previousCategories = Category.objects.filter(post=post)
-                previousCategoriesNames = [
-                    cat.category for cat in previousCategories]
-
-                # Create new categories
-                for category in categories:
-                    if category in previousCategoriesNames:
-                        previousCategoriesNames.remove(category)
-                    else:
-                        Category.objects.create(category=category, post=post)
-
-                # Remove old categories that were deleted
-                for category in previousCategoriesNames:
-                    Category.objects.get(category=category, post=post).delete()
+                categories = form.cleaned_data.get('categories')
+
+                if categories is not None:
+                    categories = categories.split()
+                    categories_to_remove = [ cat.category for cat in post.category_set.all()]
+
+                    """
+                    This implementation makes category names case-insensitive.
+                    This makes handling Category objects cleaner, albeit slightly more
+                    involved.
+                    """
+                    for category in categories:
+                        try:
+                            existing_category = Category.objects.get(category__iexact=category)     # case-insensitive lookup
+                            existing_cat_to_post = existing_category.posts.all().filter(id=post.id)
+                        except Category.DoesNotExist:
+                            existing_category = False
+                        
+                        if not existing_category:
+                            post.category_set.create(category=category)
+
+                        elif not existing_cat_to_post.exists():
+                            existing_category.posts.add(post)
+                        
+                        else:
+                            while existing_category.category in categories_to_remove:
+                                categories_to_remove.remove(existing_category.category)     # don't remove this category
+
+                    for category in categories_to_remove:
+                        cat_to_remove = Category.objects.get(category=category)
+                        if cat_to_remove.posts.count() == 1:
+                            cat_to_remove.delete()              # remove category itself
+
+                        else:
+                            cat_to_remove.posts.remove(post)    # remove ref to post
```

Comment:
I think it would be much more clear if this code block focused on only removing categories and not deleting the categories in the special case.